File: tests/src/main/java/com/github/klee0kai/test/lifecycle/base/APhone.java
package com.github.klee0kai.test.lifecycle.base;

import java.util.LinkedList;

public abstract class APhone {

    private LinkedList<APhoneLifecycle> listeners = new LinkedList<>();

    public void subscribe(APhoneLifecycle listener) {
        this.listeners.add(listener);
    }

    public void onBuy() {
        for (APhoneLifecycle lis : listeners)
            lis.onBuy();
    }

    public void onBroken() {
        for (APhoneLifecycle lis : listeners)
            lis.onBroken();
    }

    public void onDrown() {
        for (APhoneLifecycle lis : listeners)
            lis.onDrown();
    }


}

File: test_kotlin/src/main/kotlin/com/github/klee0kai/test/lifecycle/base/APhone.kt
package com.github.klee0kai.test.lifecycle.base

import java.util.*

abstract class APhone {
    private val listeners = LinkedList<APhoneLifecycle>()
    fun subscribe(listener: APhoneLifecycle) {
        listeners.add(listener)
    }

    fun onBuy() {
        for (lis in listeners) lis.onBuy()
    }

    fun onBroken() {
        for (lis in listeners) lis.onBroken()
    }

    fun onDrown() {
        for (lis in listeners) lis.onDrown()
    }
}